### PR TITLE
Bridge non-ASCII SmallStrings as native Swift Strings rather than by creating CFStrings

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by jrose-apple** [comment]
This is like three levels of wrong. 😭

**Comment by jrose-apple** [comment]
Yay, down to only 1 level of wrong!

File: test/stdlib/TestNSString.swift [hunk 1]
```diff
@@ -38,12 +38,19 @@ class TestNSString : TestNSStringSuper {
...
-    expectEqual(13, NSString(string: "\(bom)234567890123").length)
-    expectEqual(14, NSString(string: "\(bom)2345678901234").length)
+//    expectEqual(1, NSString(string: bom).length)
+//    expectEqual(4, NSString(string: "\(bom)abc").length)
+//    expectEqual(5, NSString(string: "\(bom)\(bom)abc").length)
+//    expectEqual(4, NSString(string: "a\(bom)bc").length)
+//    expectEqual(13, NSString(string: "\(bom)234567890123").length)
+//    expectEqual(14, NSString(string: "\(bom)2345678901234").length)
+    
+    expectEqual(1, (bom).length as NSString)
```
Comment (jrose-apple) [inline test/stdlib/TestNSString.swift]: Typo here…

File: test/stdlib/TestNSString.swift [hunk 2]
```diff
@@ -30,16 +30,40 @@ import FoundationBridgeObjC
...
-    }
+  func test_equalOverflow() {
+    let cyrillic = "чебурашка@ящик-с-апельсинами.рф"
+    let other = getNSStringEqualTestString()
+    print(NSStringBridgeTestEqual(cyrillic, other))
+  }
+  
+  func test_smallString_BOM() {
+    let bom = "\u{FEFF}" // U+FEFF (ZERO WIDTH NO-BREAK SPACE)
+//    expectEqual(1, NSString(string: bom).length)
```
Comment (Catfish-Man) [inline test/stdlib/TestNSString.swift]: These will continue to fail until we fix the non-smol issues, so are commented out for now

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

File: stdlib/public/core/StringBridge.swift [hunk 3]
```diff
@@ -322,12 +322,23 @@ extension String {
   public // SPI(Foundation)
   func _bridgeToObjectiveCImpl() -> AnyObject {
     if _guts.isSmall {
-      return _guts.asSmall.withUTF8 { bufPtr in
-        return _createCFString(
+      //ASCII Strings a) can bridge to tagged pointers, b) cannot contain a BOM
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift]: Space after the `//`

File: stdlib/public/core/StringBridge.swift [hunk 4]
```diff
@@ -322,12 +322,23 @@ extension String {
...
+        //We can't form a tagged pointer String, so grow to a non-small String,
+        //and bridge that instead. Also avoids CF deleting any BOM that may be
+        //present
+        var copy = self
+        copy._guts.reserveCapacity(_SmallString.capacity + 1)
+        assert(!copy._guts.isSmall)
+        return copy._bridgeToObjectiveCImpl()
       }
     }
     if _guts._object.isImmortal {
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift]: I think the logic would flow a little better as:

```swift
if _guts.isSmallASCII {
  return ...
}
if _guts.isSmall {
  return ...
}
...
```

File: test/stdlib/TestNSString.swift (hunk 2 above)
Comment (milseman) [reply test/stdlib/TestNSString.swift]: You can also split out this portion and XFAIL it instead, with the JIRA linked.

```swift
  .xfail(.always("SR-nnnn: ..."))
```

File: stdlib/public/core/StringBridge.swift [hunk 5]
```diff
@@ -322,12 +322,23 @@ extension String {
...
             bufPtr.baseAddress._unsafelyUnwrappedUnchecked,
             bufPtr.count,
             kCFStringEncodingUTF8
-        )
+          )
+        }
+      } else {
+        //We can't form a tagged pointer String, so grow to a non-small String,
+        //and bridge that instead. Also avoids CF deleting any BOM that may be
+        //present
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift]: Space after `//`

File: stdlib/public/core/StringBridge.swift [hunk 6]
```diff
@@ -322,12 +322,23 @@ extension String {
...
+        var copy = self
+        copy._guts.reserveCapacity(_SmallString.capacity + 1)
+        assert(!copy._guts.isSmall)
```
Comment (milseman) [inline stdlib/public/core/StringBridge.swift]: `.grow()` would be a tiny bit clearer, as it always grows the storage.

**Review by milseman — APPROVED** [review]
LGTM, minor improvements mentioned

File: stdlib/public/core/StringBridge.swift (hunk 4 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: Aha, didn't realize that existed

**Comment by Catfish-Man** [comment]
@swift-ci nominate

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci test

**Review by airspeedswift — APPROVED** [review]

**Pull request merged**